### ISSUE TITLE
BatchNorm vs InstanceNorm

### ISSUE THREAD

**Issue by karoly-hars** [ImagingLab/Colorizing-with-GANs#14]
Hi,
Have you tried using instance normalization instead of the usual batch normalization in either the generator and/or the discriminator? Read that it was successfully utilized for similar tasks (in pix2pix and cycle-GANs for example), but when I gave it a try in my PyTorch implementation, I achieved little to no success. Do you have an intuition on why batch normalization works better for on this specific task but not others?

**Comment by knazeri** [comment]
@karoly-hars 
Yes we did and I'm surprised if it won't have any effect! It should at least give you the same performance as batch-norm. How are your results?
We chose batch-norm over instance-norm because it was the network was converging faster. But you have to modify batch-norm a little bit. pix2pix and cycle-gan both mention instance normalization by using batch normalization with batch size of 1, which is not exactly instance-norm unless you turn off the learned statistics.
The problem with batch normalization in image generation tasks is that at inference time, the using learned affine parameters (mean and variance) might change the actual statistics of the generated image. In our implementation, we specifically turned batch normalization statistics off (`training=True` in tensorflow or `track_running_stats=False` in pytorch) which gave us better results at test time.

**Comment by karoly-hars** [comment]
It is possible, that the model just learns differently when InstanceNorm is used. The results from after 15-20 epoch, looks like there are color filters applied to the images. Maybe it would lead to the same image quality as BatchNorm, after a longer training session. 

Here are some examples:
10 epochs:
![ep10_sample](https://user-images.githubusercontent.com/39346572/51788253-0ed4d000-217c-11e9-85aa-db3fb297766e.png)
14 epochs:
![ep14_sample](https://user-images.githubusercontent.com/39346572/51788255-0f6d6680-217c-11e9-8cae-0acae62b807d.png)
17 epochs:
![ep17_sample](https://user-images.githubusercontent.com/39346572/51788257-0f6d6680-217c-11e9-8a27-65e789f39238.png)

**Comment by knazeri** [comment]
Honestly, a model with batch norm converges faster, especially when using large batch sizes! I guess it might take a while to converge. But if you are using large batch size (size > 8) then batch-normalization would give you better results. If training batch size is small (size < 8) then instance-normalization gives better results. Ans like I said before, make sure you turn off the statistics (extra parameters learned by the layer) when dealing with generative models.

**Comment by karoly-hars** [comment]
Good to know. Thank you for clarifying this.
